unittests/fork_db_tests.cpp

```diff
@@ -0,0 +1,107 @@
+   auto   bsp11a = block_state_accessor::make_unique_block_state(11, root);
+   auto     bsp12a = block_state_accessor::make_unique_block_state(12, bsp11a);
+   auto       bsp13a = block_state_accessor::make_unique_block_state(13, bsp12a);
+   auto   bsp11b = block_state_accessor::make_unique_block_state(11, root);
+   auto     bsp12b = block_state_accessor::make_unique_block_state(12, bsp11b);
+   auto       bsp13b = block_state_accessor::make_unique_block_state(13, bsp12b);
+   auto         bsp14b = block_state_accessor::make_unique_block_state(14, bsp13b);
+   auto     bsp12bb = block_state_accessor::make_unique_block_state(12, bsp11b);
+   auto       bsp13bb = block_state_accessor::make_unique_block_state(13, bsp12bb);
+   auto       bsp13bbb = block_state_accessor::make_unique_block_state(13, bsp12bb);
+   auto     bsp12bbb = block_state_accessor::make_unique_block_state(12, bsp11b);
+   auto   bsp11c = block_state_accessor::make_unique_block_state(11, root);
+   auto     bsp12c = block_state_accessor::make_unique_block_state(12, bsp11c);
+   auto       bsp13c = block_state_accessor::make_unique_block_state(13, bsp12c);
+
+   // keep track of all those added for easy verification
+   std::vector<block_state_ptr> all { bsp11a, bsp12a, bsp13a, bsp11b, bsp12b, bsp12bb, bsp12bbb, bsp13b, bsp13bb, bsp13bbb, bsp14b, bsp11c, bsp12c, bsp13c };
+
+   forkdb.reset_root(*root);
+   forkdb.add(bsp11a, mark_valid_t::no, ignore_duplicate_t::no);
+   forkdb.add(bsp11b, mark_valid_t::no, ignore_duplicate_t::no);
+   forkdb.add(bsp11c, mark_valid_t::no, ignore_duplicate_t::no);
+   forkdb.add(bsp12a, mark_valid_t::no, ignore_duplicate_t::no);
+   forkdb.add(bsp13a, mark_valid_t::no, ignore_duplicate_t::no);
+   forkdb.add(bsp12b, mark_valid_t::no, ignore_duplicate_t::no);
+   forkdb.add(bsp12bb, mark_valid_t::no, ignore_duplicate_t::no);
+   forkdb.add(bsp12bbb, mark_valid_t::no, ignore_duplicate_t::no);
+   forkdb.add(bsp12c, mark_valid_t::no, ignore_duplicate_t::no);
+   forkdb.add(bsp13b, mark_valid_t::no, ignore_duplicate_t::no);
+   forkdb.add(bsp13bb, mark_valid_t::no, ignore_duplicate_t::no);
+   forkdb.add(bsp13bbb, mark_valid_t::no, ignore_duplicate_t::no);
+   forkdb.add(bsp14b, mark_valid_t::no, ignore_duplicate_t::no);
+   forkdb.add(bsp13c, mark_valid_t::no, ignore_duplicate_t::no);
+
+   // test get_block
+   for (auto& i : all) {
+      BOOST_TEST(forkdb.get_block(i->id()) == i);
+   }
+
+   // test remove, should remove descendants
+   forkdb.remove(bsp12b->id());
+   BOOST_TEST(!forkdb.get_block(bsp12b->id()));
+   BOOST_TEST(!forkdb.get_block(bsp13b->id()));
+   BOOST_TEST(!forkdb.get_block(bsp14b->id()));
+   forkdb.add(bsp12b, mark_valid_t::no, ignore_duplicate_t::no); // will throw if already exists
+   forkdb.add(bsp13b, mark_valid_t::no, ignore_duplicate_t::no); // will throw if already exists
+   forkdb.add(bsp14b, mark_valid_t::no, ignore_duplicate_t::no); // will throw if already exists
```

Comment:
Nice test! The one in your other PR for `update_best_qc` is even better!

Reply:
Don't need the `update_best_qc` anymore, but thanks.